### PR TITLE
Add Swarm management backend

### PR DESCRIPTION
**Pull request by tonistiigi**
As described in our [ROADMAP.md](https://github.com/docker/docker/blob/master/ROADMAP.md#14cluster-capable-engine), introduce new Swarm management API endpoints relying on docker/swarmkit to deploy services. It currently vendors [docker/engine-api](docker/engine-api#265) changes.

This PR is fully backward compatible (joining a Swarm is an optional feature of the Engine, and existing commands are not impacted).

Signed-off-by: Tonis Tiigi tonistiigi@gmail.com
Signed-off-by: Victor Vieux vieux@docker.com
Signed-off-by: Daniel Nephin dnephin@docker.com

### CODE REVIEW

**Comment by thaJeztah** [comment]
Looks like some tests are failing

File: docs/reference/api/docker_remote_api_v1.24.md [hunk 1]
```diff
@@ -3306,6 +3305,1119 @@ Status Codes
...
+## 3.9 Tasks
+
+**Note:** Tasks operations require to first be part of a Swarm.
+
+### List tasks
+
+**Warning:** this endpoint is part of the Swarm management feature introduced in Docker 1.12, and
+might be subject to non backward-compatible changes.
+
+`GET /swarm/tasks`
```
Comment (aluzzardi) [inline docs/reference/api/docker_remote_api_v1.24.md]: Should be `GET /tasks`

/cc @icecrime

File: daemon/cluster/convert/swarm.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+	// Annotations
+	swarm.Spec.Name = c.Spec.Annotations.Name
+	swarm.Spec.Labels = c.Spec.Annotations.Labels
+
+	for _, policy := range c.Spec.AcceptancePolicy.Policies {
+		swarm.Spec.AcceptancePolicy.Policies = append(swarm.Spec.AcceptancePolicy.Policies, types.Policy{
+			Role:       types.NodeRole(policy.Role.String()),
+			Autoaccept: policy.Autoaccept,
+			Secret:     string(policy.Secret.Data),
```
Comment (mountkin) [inline daemon/cluster/convert/swarm.go]: Panicked here when running `docker swarm inspect`.
```
2016-06-12 15:02:57.184744 I | http: panic serving @: runtime error: invalid memory address or nil pointer dereference
goroutine 234 [running]:
net/http.(*conn).serve.func1(0xc8203c5b80)
        /usr/local/go/src/net/http/server.go:1389 +0xc1
panic(0x1a6c400, 0xc820010070)
        /usr/local/go/src/runtime/panic.go:443 +0x4e9
github.com/docker/docker/daemon/cluster/convert.SwarmFromGRPC(0xc820ee0280, 0x19, 0x9, 0xc8208c5e40, 0xc8208c5e50, 0xc8208c5e11, 0x7, 0x0, 0xc8208c5ea0, 0x2, ...)
        /go/src/github.com/docker/docker/daemon/cluster/convert/swarm.go:50 +0x267
github.com/docker/docker/daemon/cluster.(*Cluster).Inspect(0xc8200b9680, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/docker/docker/daemon/cluster/cluster.go:411 +0x23f
github.com/docker/docker/api/server/router/swarm.(*swarmRouter).inspectCluster(0xc82072a7e0, 0x7f350b7d37d8, 0xc82089db30, 0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0, 0xc82089d6e0, 0x0, 0x0)
        /go/src/github.com/docker/docker/api/server/router/swarm/cluster_routes.go:48 +0x67
github.com/docker/docker/api/server/router/swarm.(*swarmRouter).(github.com/docker/docker/api/server/router/swarm.inspectCluster)-fm(0x7f350b7d37d8, 0xc82089db30, 0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0, 0xc82089d6e0, 0x0, 0x0)
        /go/src/github.com/docker/docker/api/server/router/swarm/cluster.go:30 +0x74
github.com/docker/docker/api/server/middleware.VersionMiddleware.WrapHandler.func1(0x7f350b7d37d8, 0xc82089db30, 0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0, 0xc82089d6e0, 0x0, 0x0)
        /go/src/github.com/docker/docker/api/server/middleware/version.go:56 +0x9aa
github.com/docker/docker/api/server/middleware.UserAgentMiddleware.WrapHandler.func1(0x7f350b7d37d8, 0xc82089d980, 0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0, 0xc82089d6e0, 0x0, 0x0)
        /go/src/github.com/docker/docker/api/server/middleware/user_agent.go:45 +0x68e
github.com/docker/docker/api/server/middleware.DebugRequestMiddleware.func1(0x7f350b7c5940, 0xc8200110f8, 0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0, 0xc82089d6e0, 0x0, 0x0)
        /go/src/github.com/docker/docker/api/server/middleware/debug.go:21 +0xd4d
github.com/docker/docker/api/server.(*Server).makeHTTPHandler.func1(0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0)
        /go/src/github.com/docker/docker/api/server/server.go:139 +0x138
net/http.HandlerFunc.ServeHTTP(0xc820859820, 0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0)
        /usr/local/go/src/net/http/server.go:1618 +0x3a
github.com/gorilla/mux.(*Router).ServeHTTP(0xc82000d0e0, 0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0)
        /go/src/github.com/docker/docker/vendor/src/github.com/gorilla/mux/mux.go:98 +0x29e
github.com/docker/docker/api/server.(*routerSwapper).ServeHTTP(0xc8206c71a0, 0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0)
        /go/src/github.com/docker/docker/api/server/router_swapper.go:29 +0x72
net/http.serverHandler.ServeHTTP(0xc8202a7f80, 0x7f3508607e78, 0xc8203fa680, 0xc8214a5dc0)
        /usr/local/go/src/net/http/server.go:2081 +0x19e
net/http.(*conn).serve(0xc8203c5b80)
        /usr/local/go/src/net/http/server.go:1472 +0xf2e
created by net/http.(*Server).Serve
        /usr/local/go/src/net/http/server.go:2137 +0x44e
```

Steps to reproduce:
1. Start docker service with `/usr/bin/dockerd --debug --raw-logs`
2. `docker swarm init`
3. `docker swarm inspect`

Comment (vdemeester) [inline daemon/cluster/convert/swarm.go]: Same here :sweat: 

```
root@020824023c36:/go/src/github.com/docker/docker# docker swarm inspect
[]
An error occurred trying to connect: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.24/swarm: EOF
```

**Comment by icecrime** [comment]
LGTM :100:

**Comment by mavenugo** [comment]
🎉 🎉 LGTM

**Comment by vieux** [comment]
LGTM 💯 this is very good!

**Comment by icecrime** [comment]
API reference will be completed in a followup :+1:

**Pull request merged**